OpenRA.Game/GameRules/RulesetWatcher.cs

```diff
@@ -0,0 +1,172 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Timers;
+
+namespace OpenRA.GameRules
+{
+	public class RulesetWatcher : IDisposable
+	{
+		readonly World world;
+		readonly ModData modData;
+		readonly Dictionary<string, string> watchFiles = new();
+		readonly Timer timer;
+		readonly HashSet<string> fileQueue = new();
+		readonly FileSystemWatcher watcher;
+
+		volatile bool isEnabled;
+		bool isDisposed;
+
+		public bool IsEnabled => isEnabled;
+
+		public RulesetWatcher(World world, ModData modData)
+		{
+			this.world = world;
+			this.modData = modData;
+
+			var files = modData.Manifest.Rules
+				.Concat(modData.Manifest.Weapons)
+				.Concat(modData.Manifest.Sequences);
+			foreach (var file in files)
+			{
+				string filename;
+				using (var fs = modData.DefaultFileSystem.Open(file) as FileStream)
+					filename = fs?.Name;
+
+				if (string.IsNullOrEmpty(filename))
+					continue;
+
+				var fullPath = Path.GetFullPath(filename);
+				watchFiles[fullPath] = file;
+			}
+
+			timer = new Timer(10) { AutoReset = false };
+			timer.Elapsed += RefreshTimer;
+
+			watcher = new FileSystemWatcher(modData.Manifest.Package.Name)
+			{
+				IncludeSubdirectories = true
+			};
+			watcher.Changed += FileChanged;
+
+			foreach (var file in watchFiles.Keys.Distinct())
+			{
+				watcher.Filters.Add(Path.GetFileName(file));
+			}
+		}
+
+		public void ToggleWatching(bool toggleValue)
+		{
+			if (isDisposed)
+				throw new ObjectDisposedException(nameof(RulesetWatcher));
+
+			watcher.EnableRaisingEvents = toggleValue;
+			isEnabled = toggleValue;
+			if (toggleValue)
+				RestartTimerIfEnabled();
+			else
+			{
+				timer.Stop();
+
+				lock (fileQueue)
+					fileQueue.Clear();
+			}
+		}
+
+		void RefreshTimer(object sender, ElapsedEventArgs e)
+		{
+			if (fileQueue.Count == 0)
+			{
+				RestartTimerIfEnabled();
+				return;
+			}
+
+			List<string> copy;
+			lock (fileQueue)
+			{
+				if (fileQueue.Count == 0)
+					return;
+
+				copy = fileQueue.ToList();
+				fileQueue.Clear();
+			}
+
+			if (isDisposed || !isEnabled)
+				return;
+
+			Game.RunAfterTick(() =>
+			{
+				if (isDisposed || !isEnabled)
+					return;
+
+				var modFsFilenames = copy.Select(f => watchFiles[f]).ToArray();
+
+				var defaultRules = world.Map.Rules;
+				var rulesFiles = FindRulesetFiles(modData.Manifest.Rules, modFsFilenames).ToArray();
+				var weaponFiles = FindRulesetFiles(modData.Manifest.Weapons, modFsFilenames).ToArray();
+				var sequenceFiles = FindRulesetFiles(modData.Manifest.Sequences, modFsFilenames).ToArray();
+
+				if (rulesFiles.Length > 0)
+					TryExecute(() => defaultRules.LoadActorTraitsFromRuleFile(world, modData, rulesFiles));
+				if (weaponFiles.Length > 0)
+					TryExecute(() => defaultRules.LoadWeaponsFromFile(world, modData, weaponFiles));
+				if (sequenceFiles.Length > 0)
+					TryExecute(() => world.Map.Sequences.ReloadSequenceSetFromFiles(modData.DefaultFileSystem, sequenceFiles));
```

Comment:
Without having looked deeper into the code here, I'd be interested if this was possible to apply atomically? Or at least if not applying them atomically is ok?

i.e. if the actors rules apply okay but then the weapons fail to apply - do we end up in an okay state? The user could have rules for state A and make changes to apply state B, but we could end up in a halfway state between A and B that is neither A or B.

Maybe this is all fine, especially for a dev tool, but just throwing the thought out there really.

Reply:
Honestly, I haven't studied the code for reloading ruleset, so I can't really say. I think you're onto something, in such scenario it could put the ruleset into an inconsistent state.

So, I agree, this might be a good idea to consider.